src/Adaptive.Aeron/LogBuffer/HeaderWriter.cs

```diff
@@ -35,18 +33,15 @@ public HeaderWriter(IDirectBuffer defaultHeader)
         /// <param name="offset">     at which the header should be written. </param>
         /// <param name="length">     of the fragment including the header. </param>
         /// <param name="termId">     of the current term buffer. </param>
-        public virtual void Write(IAtomicBuffer termBuffer, int offset, int length, int termId)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe void Write(UnsafeBuffer termBuffer, int offset, int length, int termId)
         {
-            var lengthVersionFlagsType = _versionFlagsType | (-length & 0xFFFFFFFFL);
             var termOffsetSessionId = _sessionId | (uint)offset;
             var streamAndTermIds = _streamId | ((long)termId << 32);
 
-            //TODO why not just putlongvolatile?
-            termBuffer.PutLongOrdered(offset + HeaderFlyweight.FRAME_LENGTH_FIELD_OFFSET, lengthVersionFlagsType);
-            Thread.MemoryBarrier();
-
-            termBuffer.PutLong(offset + DataHeaderFlyweight.TERM_OFFSET_FIELD_OFFSET, termOffsetSessionId);
-            termBuffer.PutLong(offset + DataHeaderFlyweight.STREAM_ID_FIELD_OFFSET, streamAndTermIds);
+            *(int*)(termBuffer.BufferPointer + offset + HeaderFlyweight.VERSION_FIELD_OFFSET) = _versionFlagsType;
```

Comment:
In my final version, it is moved into TermAppender with a direct `Volatile.Write` call. Original PutLongOrdered is implemented the same way via `Volatile.Write` [here](https://github.com/AdaptiveConsulting/Aeron.NET/blob/ba58ca2e85c215f85c9c65b4e3fa9a80402eafb7/src/Adaptive.Agrona/Concurrent/UnsafeBuffer.cs#L296).